R/generate_retro_triangles.R

```diff
@@ -0,0 +1,108 @@
+#' Generate retrospective reporting triangles
+#'
+#' This function ingests a reporting triangle and the number of retrospective
+#'   reporting triangles we want to create, and iteratively generates the
+#'   reporting triangle that would have been available as of the maximum
+#'   reference time, and iteratively working backwards for
+#'   `n_history_uncertainty` snapshots
+#'
+#' @param triangle Matrix of the reporting triangle
+#'   to be used to generate retrospective nowcasts, with rows representing the
+#'   time points of reference and columns representing the delays
+#' @param n_triangles Integer indicating the number of retrospective
+#'   reporting triangles to be generated, always starting from the most
+#'   recent reference time. Default is  only generate truncated matrices
+#'   that have sufficient number of rows to generate a nowcast from, though
+#'   any number can be specified.
+#' @returns A list of `n_triangle` retrospective reporting triangle matrices
+#'   with as many rows as available given the truncation, and the same number
+#'   of columns as `triangle`
+#' @export
+#' @examples
+#' triangle <- matrix(
+#'   c(
+#'     65, 46, 21, 7,
+#'     70, 40, 20, 5,
+#'     80, 50, 10, 10,
+#'     100, 40, 31, 20,
+#'     95, 45, 21, NA,
+#'     82, 42, NA, NA,
+#'     70, NA, NA, NA
+#'   ),
+#'   nrow = 7,
+#'   byrow = TRUE
+#' )
+#'
+#' retro_rts <- generate_retro_triangles(
+#'   triangle = triangle,
+#'   n_triangles = 2
+#' )
+#' print(retro_rts[[1]])
+#' print(retro_rts[[2]])
+generate_retro_triangles <- function(
+    triangle,
+    n_triangles = nrow(triangle) - ncol(triangle)) {
+  .validate_triangle(triangle)
+  if (n_triangles > (nrow(triangle) - ncol(triangle))) {
+    cli::cli_warn(
+      message = c(
+        "Not all of the triangles generated will contain sufficient ",
+        "observations to generate a nowcast from"
+      )
+    )
+  }
+
+  # Will be able to remove this step if we require NAs in the bottom right
```

Comment:
does this mean this function only works with partially observed data and you can't take a reported rectangle and generate reporting triangles from that?

Reply:
This step actually isn't necessary anymore. If you did pass in a reported rectangle, it would still generate retrospective reporting triangles from that, with the first one assuming you hadn't observed the last row (most recent time point). 

I will remove.